qutip/tests/core/test_qobjevo.py

```diff
@@ -42,111 +42,442 @@
+
+    def test_tidyup(self):
+        "QobjEvo tidyup"
+        obj = self.qobjevos["no_args"]
+        obj *= 1e-10 * np.random.random()
+        obj.tidyup(atol=1e-8)
+        t = self._rand_t()
+        # check that the Qobj are cleaned
+        assert_equal(np.max(obj(t).full()), 0.)
+
+    def test_QobjEvo_pickle(self, form):
+        "QobjEvo pickle"
+        # used in parallel_map
+        import pickle
+        pickled = pickle.dumps(self.qobjevos[form], -1)
+        recreated = pickle.loads(pickled)
+        _assert_qobjevo_equivalent(recreated, self.qobjevos[form])
+
+    @pytest.mark.parametrize('superop', [lindblad_dissipator,
+                                         partial(lindblad_dissipator, chi=0.5),
+                                         sprepost,
+                                         lambda O1, O2: liouvillian(O1, [O2])],
+                             ids=['lindblad', 'lindblad_chi',
+                                  'sprepost', 'liouvillian'])
+    def test_superoperator_qobj(self, form, superop):
+        t = self._rand_t()
+        obj1 = self.qobjevos[form]
+        obj1_t = self.qobjevos[form](t)
+        qobj = rand_herm(self.N)
+
+        as_qevo = superop(obj1, qobj)(t)
+        as_qobj = superop(obj1_t, qobj)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+        as_qevo = superop(qobj, obj1)(t)
+        as_qobj = superop(qobj, obj1_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    @pytest.mark.parametrize('superop', [lindblad_dissipator,
+                                         partial(lindblad_dissipator, chi=0.5),
+                                         sprepost,
+                                         lambda O1, O2: liouvillian(O1, [O2])],
+                             ids=['lindblad', 'lindblad_chi',
+                                  'sprepost', 'liouvillian'])
+    def test_superoperator(self, form, extra_form, superop):
+        t = self._rand_t()
+        obj1 = self.qobjevos[form]
+        obj1_t = self.qobjevos[form](t)
+        obj2 = self.qobjevos[extra_form]
+        obj2_t = self.qobjevos[extra_form](t)
+
+        as_qevo = superop(obj1, obj2)(t)
+        as_qobj = superop(obj1_t, obj2_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    def test_shift(self, form):
+        t = self._rand_t() / 2
+        dt = self._rand_t() / 2
+        obj = self.qobjevos[form]
+        shifted = obj._shift()
+        _assert_qobj_almost_eq(obj(t + dt), shifted(t, args={"_t0": dt}))
+
+    def test_QobjEvo_apply(self):
+        "QobjEvo apply"
+        obj = self.qobjevos["no_args"]
+        transposed = obj.apply(_trans)
+        _assert_qobjevo_equivalent(transposed, obj.trans())
+
+    def test_mul_vec(self, form):
+        "QobjEvo mul_vec"
+        t = self._rand_t()
+        N = self.N
+        vec = np.arange(N)*.5+.5j
+        op = self.qobjevos[form]
+        assert np.allclose(op(t).full() @ vec,
+                           op.mul_vec(t, vec))
```

Comment:
this is one scenario where `pytest`'s bare assert isn't as good as `np.testing.assert_allclose` - currently `pytest` doesn't treat `numpy` arrays specially in the introspection